src/zjs_gpio.c

```diff
@@ -420,11 +490,398 @@ static ZJS_DECL_FUNC(zjs_gpio_open_async)
     return ZJS_CHAIN_FUNC_ARGS(zjs_gpio_open, true);
 }
 
+#ifdef ZJS_GPIO_MOCK
+// Sample structure of mock root object:
+// var obj = {
+//     "GPIO_0": {
+//         "pin1": {
+//             "state": <boolean>,
+//             "flags": <uint32>,
+//             "wired": <array>  // pins this pin is wired to
+//         },
+//         "pin2": <object>
+//     },
+//     "GPIO_1": <object>,
+//     "GPIO_2": <object>,
+// }
+
+// mock helpers
+static jerry_value_t get_pin(jerry_value_t port, unsigned int pin)
+{
+    char name[15];
+    sprintf(name, "pin%u\n", pin);
+    return zjs_get_property(port, name);
+}
+
+static jerry_value_t ensure_pin(jerry_value_t port, unsigned int pin)
+{
+    char name[15];
+    sprintf(name, "pin%u\n", pin);
+    jerry_value_t pin_obj = zjs_get_property(port, name);
+    if (!jerry_value_is_object(pin_obj)) {
+        pin_obj = jerry_create_object();
+        zjs_set_property(port, name, pin_obj);
+    }
+    return pin_obj;
+}
+
+struct value_match_data {
+    jerry_value_t name;
+    jerry_value_t value;
+};
+
+static bool value_match(const jerry_value_t name, const jerry_value_t value,
+                        void *user_data)
+{
+    struct value_match_data *data = (struct value_match_data *)user_data;
+    if (value == data->value) {
+        // found match, stop processing and return name
+        data->name = name;
+        return false;
+    }
+    return true;
+}
+
+static jerry_value_t find_property_with_value(jerry_value_t obj,
```

Comment:
This would be a good function for zjs_util as well as push_array(). Keep it here for now but I think OCF could use it.